cli/tests/common/test_environment.rs

```diff
@@ -338,6 +365,71 @@ impl TestEnvironment {
     }
 }
 
+/// Command output and exit status to be displayed in normalized form.
+#[derive(Clone)]
+pub struct CommandOutput {
+    pub stdout: CommandOutputString,
+    pub stderr: CommandOutputString,
+    pub status: ExitStatus,
+}
+
+impl CommandOutput {
+    /// Normalizes Windows directory separator to slash.
+    #[must_use]
+    pub fn normalize_backslash(self) -> Self {
+        CommandOutput {
+            stdout: self.stdout.normalize_backslash(),
+            stderr: self.stderr.normalize_backslash(),
+            status: self.status,
+        }
+    }
+
+    /// Removes the last line (such as platform-specific error message) from the
+    /// normalized stderr text.
+    #[must_use]
+    pub fn strip_stderr_last_line(self) -> Self {
+        CommandOutput {
+            stdout: self.stdout,
+            stderr: self.stderr.strip_last_line(),
+            status: self.status,
+        }
+    }
+
+    /// Ensures that the command exits with success status.
+    #[track_caller]
+    pub fn success(self) -> Self {
+        assert!(self.status.success(), "{self}");
+        self
+    }
+}
+
+impl Display for CommandOutput {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let CommandOutput {
+            stdout,
+            stderr,
+            status,
+        } = self;
+        if !stdout.is_empty() {
+            writeln!(f, "------- stdout -------")?;
+            write!(f, "{stdout}")?;
+        }
+        if !stderr.is_empty() {
+            writeln!(f, "------- stderr -------")?;
+            write!(f, "{stderr}")?;
+        }
+        if !status.success() {
+            if let Some(code) = status.code() {
+                writeln!(f, "[exit status: {code}]")?;
+            } else {
+                // On Windows, this would print "exit code: N".
```

Comment:
nit: do you mean "code" -> "status"?

Reply:
No. The output is platform-dependent, so we normalize it by not using the default formatting.

Reply:
Sorry, I still don't follow. `{status}` is the default formatting, right? And what does "this" in "this would print" refer to?

Reply:
I mean `{status}` would print `exit code: N` on Windows, so don't remove the  `if` block above.

Reply:
Ah, I see. Mayne something like this then?

```rust
// If there is an exit code, `{status}` would get rendered as "exit code: N"
// on Windows, so we render it ourselves for compatibility.
if let Some(code) = status.code() {
    writeln!(f, "[exit status: {code}]")?;
} else {
    writeln!(f, "[{status}]")?;
}
```

Reply:
If <https://github.com/jj-vcs/jj/pull/5749#discussion_r1962702592> is correct, does the first branch of the `if` need to print `[exit code...` instead of `[exit status...`?

Reply:
No. I tend to use the Unix convention over Windows one.

Reply:
I'm not sure I follow, but the tests seem to pass on both Unix on Windows, so I don't need to delay you further.

Reply:
If we simply do `writeln!(f, "[{status}]")`, the test would fail. (I haven't tried, but I know that `status` is rendered differently on Windows.)

Reply:
@ilyagr, if there is an exit code (`status.code()` is `Some`), then `{status}` gets rendered as "exit status: N" on Linux and "exit code: N" on Windows.

Reply:
I didn't notice the `normalize_exit_status` function. I think the comment here is still confusing, but it's not a big deal; I might fix it up if I remember.

I actually like Yuya's original version of having the comment in the `if` branch, but I'd say something like

> This branch gets triggered on Windows, where it prints "status code: N". The difference with Unix gets normalized away in `normalize_exit_status`.

Reply:
I don't think that's correct. The branch does not depend on OS, IIUC. The lack of exit code instead signals (pun not intended) that the process was killed by a signal.

Reply:
Yeah, the `else` path is for unusual cases in which it's okay to emit platform-dependent output.